splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/upgrade/UpgradeUtils.java

```diff
@@ -0,0 +1,72 @@
+import com.splicemachine.si.impl.driver.SIDriver;
+import com.splicemachine.storage.*;
+import com.splicemachine.utils.SpliceLogUtils;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.util.List;
+
+public class UpgradeUtils {
+
+    protected static final Logger LOG = Logger.getLogger(UpgradeUtils.class);
+
+    public static void initializeConglomerateSITable(TransactionController tc) throws IOException {
+        SIDriver driver = SIDriver.driver();
+        int rowsRewritten = 0;
+        try {
+            EntryDecoder entryDecoder = new EntryDecoder();
+            TxnView txn = ((SpliceTransactionManager) tc).getActiveStateTxn();
+            BitSet fields = new BitSet();
+            fields.set(0);
+            EntryEncoder entryEncoder = EntryEncoder.create(SpliceKryoRegistry.getInstance(), 1, fields, null, null, null);
+            try (Partition sourceTable = driver.getTableFactory().getTable(SIConfigurations.CONGLOMERATE_TABLE_NAME);
+                 Partition destTable = driver.getTableFactory().getTable(HBaseConfiguration.CONGLOMERATE_SI_TABLE_NAME)) {
+                try (DataScanner scanner = sourceTable.openScanner(driver.baseOperationFactory().newScan())) {
+                    while (true) {
+                        List<DataCell> cells = scanner.next(0);
+                        if (cells.isEmpty())
+                            break;
+                        for (DataCell cell : cells) {
+                            CellType type = cell.dataType();
+                            if (type == CellType.USER_DATA) {
+                                byte[] key = cell.key();
+                                byte[] data = cell.value();
+                                entryDecoder.set(data);
+                                MultiFieldDecoder decoder = entryDecoder.getEntryDecoder();
+                                byte[] nextRaw = decoder.decodeNextBytesUnsorted();
+                                Conglomerate conglomerate = DerbyBytesUtil.fromBytesUnsafe(nextRaw);
+                                byte[] conglomData = DerbyBytesUtil.toBytes(conglomerate);
+                                DataPut put = driver.getOperationFactory().newDataPut(txn, key);
+                                MultiFieldEncoder encoder = entryEncoder.getEntryEncoder();
+                                encoder.reset();
+                                encoder.encodeNextUnsorted(conglomData);
+                                put.addCell(SIConstants.DEFAULT_FAMILY_BYTES, SIConstants.PACKED_COLUMN_BYTES, entryEncoder.encode());
+                                destTable.put(put);
+                                rowsRewritten++;
+                            }
+                        }
+                    }
+                }
+            }
+            SpliceLogUtils.info(LOG, "Wrote %d rows to SPLICE_CONGLOMERATE_SI.", rowsRewritten);
+        }
+        catch (IOException | ClassNotFoundException | StandardException e) {
```

Comment:
If there was a failure to add all of the entries, maybe we should drop the conglomerate SI table or do something to force the upgrade to try again.  Otherwise we may be in a state where not all entries were copied over properly.  But I see in readConglomerate() it is anticipating the error case and reading from the old table, which may be acceptable too.  Maybe we should use destTable.writeBatch instead of destTable.put in case the very first entry fails that won't mean nothing is copied to the new table.

Reply:
Yes, if there is an upgrade failure, SPLICE_CONGLOMERATE_SI should be dropped and upgrade should abort